ouroboros-consensus/src/Ouroboros/Storage/VolatileDB/Impl.hs

```diff
@@ -472,106 +476,143 @@ mkInternalStateDB :: (HasCallStack, MonadThrow m, MonadCatch m, Ord blockId)
                   -> Parser e m blockId
                   -> Int
                   -> m (InternalState blockId h)
-mkInternalStateDB hasFS@HasFS{..} err parser maxBlocksPerFile = wrapFsError hasFsErr err $ do
-    allFiles <- do
-        createDirectoryIfMissing True dir
-        Set.map toFsPath <$> listDirectory dir
+mkInternalStateDB hasFS@HasFS{..} err parser maxBlocksPerFile =
+  wrapFsError hasFsErr err $ do
+    createDirectoryIfMissing True dir
+    allFiles <- Set.map toFsPath <$> listDirectory dir
     mkInternalState hasFS err parser maxBlocksPerFile allFiles
   where
     dir = mkFsPath []
 
     toFsPath :: String -> FsPath
     toFsPath file = mkFsPath [file]
 
+-- | This function may create a new
 mkInternalState :: forall blockId m h e. (HasCallStack, MonadCatch m, Ord blockId)
                 => HasFS m h
                 -> ErrorHandling (VolatileDBError blockId) m
                 -> Parser e m blockId
                 -> Int
                 -> Set FsPath
                 -> m (InternalState blockId h)
-mkInternalState hasFS@HasFS{..} err parser n files = wrapFsError hasFsErr err $ do
-    lastFd <- findNextFd err files
-    let
+mkInternalState hasFS err parser n files =
+  wrapFsError (hasFsErr hasFS) err $ do
+    lastFd <- findLastFdM err files
+    mkInternalState' hasFS err parser n files lastFd
+
+mkInternalState' :: forall blockId m h e. (HasCallStack, MonadCatch m, Ord blockId)
+                 => HasFS m h
+                 -> ErrorHandling (VolatileDBError blockId) m
+                 -> Parser e m blockId
+                 -> Int
+                 -> Set FsPath
+                 -> Maybe FileId
+                 -> m (InternalState blockId h)
+mkInternalState' hasFS err parser n files lastFd =
+    go Index.empty Map.empty Map.empty Nothing [] (Set.toList files)
+    where
+        newFileInfo mp newIndex =
+            ( newFile
+            , newIndex + (1 :: Int)
+            , Index.insert newFile FileInfo.empty mp
+            , 0 )
+            where
+                newFile = filePath newIndex
+
+        errorMsg = concat
+            [ "Volatile db invariant violation:"
+            , "A file was found with less than "
+            , show n
+            , " blocks, but there are no files parsed."]
+
+        truncateOnError Nothing _ _ = return ()
+        truncateOnError (Just _) file offset =
+            -- the handle of the parser is closed at this point. We need
+            -- to reOpen the file in AppendMode now (parser opens with
+            -- ReadMode).
```

Comment:
Capitalise the sentence and make it clear what is code by using Haddock markup (`'..'` or `@..@` for non-linked things):
```haskell
-- The handle of the parser is closed at this point. We need
-- to reopen the file in 'AppendMode' now (parser opens with
-- 'ReadMode').